File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
    vector<vector<int>>ans;
    vector<int> currans;
public:
    void combinationSum(vector<int>& candidates, int target,int i=0) {
        if(target==0)
        {
            ans.push_back(currans);
            return;
        }
        for(int j=i;j<candidates.size();j++)
        {
            if(i!=j && candidates[j]==candidates[j-1])
                continue;
            if(target-candidates[j]<0)
                break;
            currans.push_back(candidates[j]);
            combinationSum(candidates,target-candidates[j],j+1);
            currans.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        combinationSum(candidates,target,0);
        return ans;
    }
};

File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        rv = []
        candidates.sort()
        
        def add(l,pos):
            if sum(l)<target:
                for i in range(pos,len(candidates)):
                    if i!=pos and candidates[i-1]==candidates[i]:
                        continue
                    add(l+[candidates[i]],i+1)
            elif(sum(l)==target):
                rv.append(l)
        
        add([],0)
        return rv